src/nafill.c

```diff
@@ -0,0 +1,194 @@
+}
+
+SEXP nafillR(SEXP obj, SEXP type, SEXP fill, SEXP inplace, SEXP cols, SEXP verbose) {
+  int protecti=0;
+  if (!isLogical(verbose) || length(verbose)!=1 || LOGICAL(verbose)[0]==NA_LOGICAL)
+    error("verbose must be TRUE or FALSE");
+  bool bverbose = LOGICAL(verbose)[0];
+
+  if (!xlength(obj)) return(obj);
+
+  bool binplace = LOGICAL(inplace)[0];
+  SEXP x = R_NilValue;
+  if (isVectorAtomic(obj)) {
+    if (binplace) {
+      error("'x' argument is atomic vector, in-place update is supported only for list/data.table");
+    } else if (!(isReal(obj) || isInteger(obj))) {
+      error("'x' argument must be numeric type, or list/data.table of numeric types");
+    }
+    x = PROTECT(allocVector(VECSXP, 1)); protecti++; // wrap into list
+    SET_VECTOR_ELT(x, 0, obj);
+  } else {
+    SEXP ricols = PROTECT(colnamesInt(obj, cols)); protecti++; // nafill cols=NULL which turns into seq_along(obj)
+    x = PROTECT(allocVector(VECSXP, length(ricols))); protecti++;
+    int *icols = INTEGER(ricols);
+    for (int i=0; i<length(ricols); i++) {
+      if (!(isReal(VECTOR_ELT(obj, icols[i]-1)) || isInteger(VECTOR_ELT(obj, icols[i]-1)))) {
+        error("'x' argument must be numeric type, or list/data.table of numeric types");
+      }
+      SET_VECTOR_ELT(x, i, VECTOR_ELT(obj, icols[i]-1));
+    }
+  }
+  R_len_t nx = length(x);
+
+  double* dx[nx];
+  int32_t* ix[nx];
+  uint_fast64_t inx[nx];
+  SEXP ans = R_NilValue;
+  ans_t vans[nx];
+  for (R_len_t i=0; i<nx; i++) {
+    inx[i] = xlength(VECTOR_ELT(x, i));
+    dx[i] = REAL(VECTOR_ELT(x, i));
+    ix[i] = INTEGER(VECTOR_ELT(x, i));
+  }
+  if (!binplace) {
+    ans = PROTECT(allocVector(VECSXP, nx)); protecti++;
+    for (R_len_t i=0; i<nx; i++) {
+      SET_VECTOR_ELT(ans, i, allocVector(TYPEOF(VECTOR_ELT(x, i)), inx[i]));
+      vans[i] = ((ans_t) { .dbl_v=REAL(VECTOR_ELT(ans, i)), .int_v=INTEGER(VECTOR_ELT(ans, i)), .status=0, .message={"\0","\0","\0","\0"} });
+    }
+  } else {
+    for (R_len_t i=0; i<nx; i++) {
+      vans[i] = ((ans_t) { .dbl_v=dx[i], .int_v=ix[i], .status=0, .message={"\0","\0","\0","\0"} });
+    }
+  }
+
+  unsigned int itype;
+  if (!strcmp(CHAR(STRING_ELT(type, 0)), "const")) itype = 0;
+  else if (!strcmp(CHAR(STRING_ELT(type, 0)), "locf")) itype = 1;
+  else if (!strcmp(CHAR(STRING_ELT(type, 0)), "nocb")) itype = 2;
+  else error("Internal error: invalid type argument in nafillR function, should have been caught before. please report to data.table issue tracker."); // # nocov
+
+  if (itype==0 && length(fill)!=1)
+    error("fill must be a vector of length 1");
+
+  double dfill=NA_REAL;
+  int32_t ifill=NA_INTEGER;
+  if (itype==0) {
+    if (isInteger(fill)) {
+      ifill = INTEGER(fill)[0];
+      dfill = INTEGER(fill)[0]==NA_INTEGER ? NA_REAL : (double)INTEGER(fill)[0];
+    } else if (isReal(fill)) {
+      ifill = ISNA(REAL(fill)[0]) ? NA_INTEGER : (int32_t)REAL(fill)[0];
```

Comment:
@mattdowle is this type of casting safe? and similar in line 152. Should I always use `coerceVector` in such cases?

Reply:
it looks that `coerceVector` is not necessary: https://github.com/Rdatatable/data.table/pull/3627/commits/638382355a08bb9fe2d4b7927702548695e74e2a#diff-74714f9d6425434f378360350d1a4c78R41